src/main/java/org/terasology/gookeeper/system/EconomySystem.java

```diff
@@ -75,32 +93,120 @@
     }
 
+    @ReceiveEvent
+    public void onPlayerSpawn(OnPlayerSpawnedEvent event, EntityRef player, InventoryComponent inventory) {
+        if (player.hasComponent(EconomyComponent.class)) {
+            nuiManager.getHUD().addHUDElement("PlayerHud");
+        }
+    }
+
     @Override
     public void shutdown() {
     }
 
     @Override
     public void update(float delta) {
-        // Have to add the wallet hud this way since adding in within initialise() results in the economy component to be null
-        if (!setHud) {
-            for (EntityRef wallet : entityManager.getEntitiesWith(EconomyComponent.class)) {
-                EconomyComponent economyComponent = wallet.getComponent(EconomyComponent.class);
-                if (economyComponent != null) {
-                    nuiManager.getHUD().addHUDElement("PlayerHud");
-                    setHud = true;
+    }
+
+    /**
+     * This command can be used to purchase utility blocks (such as pen blocks, visit blocks, etc.)
+     *
+     * @param client
+     * @param itemPrefabName
+     * @param amount
+     * @param shapeUriParam
+     * @return
+     */
+    @Command(shortDescription = "Purchase utility blocks",
+            requiredPermission = PermissionManager.NO_PERMISSION)
+    public String purchase(@Sender EntityRef client,
+                           @CommandParam("prefabId or blockName") String itemPrefabName,
+                           @CommandParam(value = "amount", required = false) Integer amount,
+                           @CommandParam(value = "blockShapeName", required = false) String shapeUriParam) {
+        EntityRef player = client.getComponent(ClientComponent.class).character;
+        EconomyComponent economyComponent = player.getComponent(EconomyComponent.class);
+
+        int quantityParam = amount != null ? amount : 16;
+
+        if (economyComponent.playerWalletCredit - (quantityParam * Assets.getPrefab(itemPrefabName + "Fenced").get().getComponent(PurchasableComponent.class).basePrice) > 0f) {
+            String message = blockCommands.giveBlock(client, itemPrefabName, amount, shapeUriParam);
+            if (message != null) {
+                economyComponent.playerWalletCredit -= quantityParam * Assets.getPrefab(itemPrefabName + "Fenced").get().getComponent(PurchasableComponent.class).basePrice;
+                player.saveComponent(economyComponent);
+                return "Successfully purchased " + itemPrefabName;
+            } else {
+                return "Couldn't find requested block.";
+            }
+        } else {
+            return "You dont have sufficient balance to purchase.";
+        }
+    }
+
+    /**
+     * This command can be used to purchase utility blocks (such as pen blocks, visit blocks, etc.)
+     *
+     * @param client
+     * @param itemPrefabName
+     * @param amount
+     * @param shapeUriParam
+     * @return
+     */
+    @Command(shortDescription = "Purchase utility blocks",
```

Comment:
The description is "Purchase utility blocks" but the command is `upgrade`? Seems like one of the two is mismatched